das/manager.go

```diff
@@ -0,0 +1,187 @@
+package das
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+// samplingManager coordinates sampling workers and stores current state
+type samplingManager struct {
+	concurrency int
+	fetchFn     fetchFn
+	storeState  storeFn
+	state       *state
+
+	catchUpDone   int32         // indicates if all headers are sampled
+	catchUpDoneCh chan struct{} // blocks until all headers are sampled
+
+	workersWg     sync.WaitGroup
+	jobsCh        chan uint64 // fan-out jobs to workers
+	resultCh      chan result // fan-in sampling results from worker to coordinator
+	updMaxCh      chan uint64 // signals to update max known header height
+	storeCh       chan state  // communicates with backgroundStore routine
+	storeInterval time.Duration
+
+	coordinatorDone chan struct{}
+	cancel          context.CancelFunc
+}
+
+type fetchFn func(context.Context, uint64) error
+type storeFn func(context.Context, state)
+
+func newSamplingManager(concurrency int,
+	bufferSize int,
+	storeInterval time.Duration,
+	fetch fetchFn,
+	storeFn storeFn) *samplingManager {
+	return &samplingManager{
+		concurrency:     concurrency,
+		fetchFn:         fetch,
+		storeState:      storeFn,
+		jobsCh:          make(chan uint64, bufferSize),
+		resultCh:        make(chan result),
+		updMaxCh:        make(chan uint64),
+		storeCh:         make(chan state),
+		storeInterval:   storeInterval,
+		coordinatorDone: make(chan struct{}),
+		catchUpDoneCh:   make(chan struct{}),
+	}
+}
+
+func (sm *samplingManager) run(ctx context.Context, ch checkpoint) {
+	ctx, sm.cancel = context.WithCancel(ctx)
+	sm.state = ch.toSamplingState()
+
+	go sm.runCoordinator(ctx)
+	if sm.storeInterval > 0 {
+		// run store routine only when storeInterval is specified
+		go sm.runBackgroundStore(ctx, sm.storeInterval)
+	}
+
+	// launch workers
+	for i := 0; i < sm.concurrency; i++ {
+		sm.workersWg.Add(1)
+		go func() {
+			defer sm.workersWg.Done()
+			runWorker(ctx, sm.jobsCh, sm.resultCh, sm.fetchFn)
+		}()
+	}
+}
+
+func (sm *samplingManager) runCoordinator(ctx context.Context) {
+	jobsCh := sm.jobsCh
+	noop := make(chan uint64)
+
+	var next uint64
+	var done bool
+	for {
+		sm.updateStats()
+
+		if next, done = sm.state.nextHeight(); done {
+			// if nothing to sample, don't send job to workers
+			jobsCh = noop
+		}
+
+		select {
+		case jobsCh <- next:
+			sm.state.setBusy(next)
+		case max := <-sm.updMaxCh:
+			if sm.state.updateMaxKnown(max) && done {
+				// if found new headers to sample and jobsCh was locked - unblock it
+				jobsCh = sm.jobsCh
+			}
+		case res := <-sm.resultCh:
+			sm.state.handleResult(res)
+		case sm.storeCh <- *sm.state:
```

Comment:
I am a bit concerned about the copy here and wondering if it gets copied on each loop/try to write to `sm.storeCh` even when its blocked. Basically, there might be some idle copies that we should avoid, but this has to be verified.

Reply:
We can investigate this or just interpolate the storing ticker into the runCoordinator. There is no big reason to keep it separated

Reply:
Considering that state is only used inside samplingManager, you can store it as value(without asterisk) and writing ot the storeCh will do copying automatically

Reply:
good point, changed it to be value field instead of pointer